File: raspberry/SmartHome/app/src/main/java/smarthome/raspberry/MainActivity.java
package smarthome.raspberry;

import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.util.Log;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;


import smarthome.library.datalibrary.store.firestore.FirestoreSmartHomeStorage;
import smarthome.raspberry.auth.GoogleSignInActivity;
import smarthome.raspberry.server.StoppableServer;
import smarthome.raspberry.server.UdpServer;
import smarthome.raspberry.server.WebServer;
import smarthome.raspberry.utils.HomeController;
import smarthome.raspberry.utils.SharedPreferencesHelper;

import static smarthome.raspberry.model.device.constants.Constants.RC_SIGN_IN;


public class MainActivity extends Activity {

    public static final String TAG = MainActivity.class.getSimpleName();
    public static final boolean DEBUG = BuildConfig.DEBUG;
    private WebServer httpServer;
    private UdpServer udpServer;
    private FirebaseAuth mAuth;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        if (DEBUG) Log.d(TAG, "onCreate");

        httpServer = new WebServer();
        udpServer = new UdpServer();
    }

    @Override
    protected void onStart() {
        super.onStart();
        startServer(httpServer);
        startServer(udpServer);
    }

    @Override
    protected void onResume() {
        super.onResume();
        auth();
    }

    @Override
    protected void onStop() {
        super.onStop();
        stopServer(httpServer);
        stopServer(udpServer);
    }

    private void auth() {
        // check for auth
        mAuth = FirebaseAuth.getInstance();

        if (mAuth.getCurrentUser() == null)
            startActivity(new Intent(this, GoogleSignInActivity.class));
    }

    private void startServer(StoppableServer server) {
        if (server == null) return;
        server.startServer();
    }

    private void stopServer(StoppableServer server) {
        if (server == null) return;
        server.stopServer();
    }
}

File: raspberry/SmartHome/app/src/main/java/smarthome/raspberry/MainActivity.kt
package smarthome.raspberry

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.util.Log

import kotlinx.coroutines.*


import smarthome.raspberry.auth.GoogleSignInActivity
import smarthome.raspberry.domain.usecases.AuthUseCase
import smarthome.raspberry.domain.usecases.DevicesUseCase
import smarthome.raspberry.domain.usecases.HomeUseCase
import smarthome.raspberry.model.SmartHomeRepository
import smarthome.raspberry.model.listeners.RepoInitListener
import smarthome.raspberry.arduinodevices.server.UdpServer
import smarthome.raspberry.arduinodevices.server.WebServer
import smarthome.raspberry.service.DeviceObserver

private val TAG = MainActivity::class.java.simpleName
private val DEBUG = BuildConfig.DEBUG
private val AUTH_ACTIVITY_REQUEST_CODE = 12312

class MainActivity : Activity(), RepoInitListener {
    private var httpServer: WebServer? = null
    private var udpServer: UdpServer? = null
    private val authUseCase: AuthUseCase = TODO()
    private val devicesUseCase: DevicesUseCase = TODO()
    private val homeUseCase: HomeUseCase = TODO()


    val deviceObserver: DeviceObserver = DeviceObserver.getInstance()


    private val job = Job()
    private val uiScope = CoroutineScope(Dispatchers.Main + job)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        if (DEBUG) Log.d(TAG, "onCreate")

        httpServer = WebServer()
        udpServer = UdpServer()
    }

    override fun onStart() {
        super.onStart()
        httpServer?.startServer()
        udpServer?.startServer()
    }

    override fun onResume() {
        super.onResume()

        if (!authUseCase.isAuthenticated()) {
            startActivity(Intent(this, GoogleSignInActivity::class.java))
        } else {
            uiScope.launch { homeUseCase.start() }
        }
    }

    override fun onStop() {
        super.onStop()
        httpServer?.stopServer()
        udpServer?.stopServer()
    }

    override fun onDestroy() {
        super.onDestroy()
        job.cancel()
    }

    override suspend fun onInitializationComplete() {
        SmartHomeRepository.listenForCloudChanges()
        SmartHomeRepository.subscribeToMessageQueue()
        deviceObserver.start()
    }
}
